File: src/infrastructure/waveform_cache.py
from __future__ import annotations

import hashlib
import json
import os
import shutil
import sys
import threading
import time
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Callable, Optional

from src.domain.services.vcd_index_format import DATA_MAGIC, INDEX_VERSION, validate_manifest
from src.domain.services.vcd_index_service import build_vcd_index


DEFAULT_MAX_BYTES = 4 * 1024**3
SAMPLE_BYTES = 64 * 1024
DEFAULT_STALE_LOCK_SECONDS = 30.0


@dataclass(frozen=True)
class SourceFingerprint:
    key: str
    normalized_path: str
    size: int
    mtime_ns: int
    head_sha256: str
    tail_sha256: str
    index_version: int


def waveform_cache_root() -> Path:
    if sys.platform == "win32":
        base = Path(os.environ.get("LOCALAPPDATA", Path.home() / "AppData" / "Local"))
        return base / "VeriFlow" / "waveform-cache"
    if sys.platform == "darwin":
        return Path.home() / "Library" / "Caches" / "VeriFlow" / "waveform-cache"
    base = Path(os.environ.get("XDG_CACHE_HOME", Path.home() / ".cache"))
    return base / "veriflow" / "waveform-cache"


def _normalized_source_path(source: Path) -> str:
    normalized = str(source.resolve()).replace("\\", "/")
    return normalized.lower() if os.name == "nt" else normalized


def _sample_hashes(source: Path, size: int) -> tuple[str, str]:
    with source.open("rb") as handle:
        head = handle.read(SAMPLE_BYTES)
        handle.seek(max(0, size - SAMPLE_BYTES))
        tail = handle.read(SAMPLE_BYTES)
    return hashlib.sha256(head).hexdigest(), hashlib.sha256(tail).hexdigest()


def source_fingerprint(source: Path) -> SourceFingerprint:
    source = Path(source).resolve()
    stat = source.stat()
    normalized_path = _normalized_source_path(source)
    head_sha256, tail_sha256 = _sample_hashes(source, stat.st_size)
    digest = hashlib.sha256()
    digest.update(b"VFI-CACHE-1\0")
    digest.update(normalized_path.encode("utf-8"))
    digest.update(b"\0")
    digest.update(str(stat.st_size).encode("ascii"))
    digest.update(b"\0")
    digest.update(str(stat.st_mtime_ns).encode("ascii"))
    digest.update(b"\0")
    digest.update(bytes.fromhex(head_sha256))
    digest.update(bytes.fromhex(tail_sha256))
    digest.update(str(INDEX_VERSION).encode("ascii"))
    return SourceFingerprint(
        key=digest.hexdigest(),
        normalized_path=normalized_path,
        size=stat.st_size,
        mtime_ns=stat.st_mtime_ns,
        head_sha256=head_sha256,
        tail_sha256=tail_sha256,
        index_version=INDEX_VERSION,
    )


def _pid_alive(pid: int) -> bool:
    if pid <= 0:
        return False
    try:
        os.kill(pid, 0)
    except PermissionError:
        return True
    except OSError:
        return False
    return True


class _CacheLock:
    def __init__(self, path: Path, stale_seconds: float) -> None:
        self.path = path
        self.stale_seconds = stale_seconds
        self.token = uuid.uuid4().hex
        self._stop = threading.Event()
        self._thread: Optional[threading.Thread] = None
        self._acquired = False

    def _payload(self) -> dict:
        now_ms = time.time_ns() // 1_000_000
        return {
            "pid": os.getpid(),
            "createdAtMs": now_ms,
            "heartbeatMs": now_ms,
            "token": self.token,
        }

    def _write_initial(self) -> None:
        descriptor = os.open(self.path, os.O_CREAT | os.O_EXCL | os.O_WRONLY)
        try:
            data = json.dumps(self._payload(), separators=(",", ":")).encode("utf-8")
            os.write(descriptor, data)
            os.fsync(descriptor)
        finally:
            os.close(descriptor)

    def _write_heartbeat(self) -> None:
        temporary = self.path.with_name(f"{self.path.name}.heartbeat.{self.token}")
        try:
            temporary.write_text(
                json.dumps(self._payload(), separators=(",", ":")),
                encoding="utf-8",
            )
            current = json.loads(self.path.read_text(encoding="utf-8"))
            if current.get("token") == self.token:
                os.replace(temporary, self.path)
        except (OSError, ValueError, json.JSONDecodeError):
            pass
        finally:
            temporary.unlink(missing_ok=True)

    def _heartbeat_loop(self) -> None:
        interval = max(0.1, self.stale_seconds / 3)
        while not self._stop.wait(interval):
            self._write_heartbeat()

    def _reclaim_if_stale(self) -> bool:
        try:
            original = self.path.read_bytes()
            payload = json.loads(original.decode("utf-8"))
            heartbeat_ms = int(payload.get("heartbeatMs", 0))
            pid = int(payload.get("pid", 0))
        except (OSError, ValueError, json.JSONDecodeError):
            return False
        stale = not _pid_alive(pid) or (
            time.time_ns() // 1_000_000 - heartbeat_ms > self.stale_seconds * 1000
        )
        if not stale:
            return False
        try:
            if self.path.read_bytes() != original:
                return False
            self.path.unlink()
            return True
        except OSError:
            return False

    def acquire(self) -> bool:
        for _attempt in range(2):
            try:
                self._write_initial()
                self._acquired = True
                self._thread = threading.Thread(target=self._heartbeat_loop, daemon=True)
                self._thread.start()
                return True
            except FileExistsError:
                if not self._reclaim_if_stale():
                    return False
        return False

    def release(self) -> None:
        if not self._acquired:
            return
        self._stop.set()
        if self._thread is not None:
            self._thread.join(timeout=1.0)
        try:
            payload = json.loads(self.path.read_text(encoding="utf-8"))
            if payload.get("token") == self.token:
                self.path.unlink(missing_ok=True)
        except (OSError, ValueError, json.JSONDecodeError):
            pass
        self._acquired = False


class WaveformCache:
    def __init__(
        self,
        *,
        root: Optional[Path] = None,
        max_bytes: int = DEFAULT_MAX_BYTES,
        builder: Callable = build_vcd_index,
        stale_lock_seconds: float = DEFAULT_STALE_LOCK_SECONDS,
        wait_seconds: float = 0.1,
    ) -> None:
        if max_bytes <= 0:
            raise ValueError("waveform cache size must be positive")
        self.root = Path(root) if root is not None else waveform_cache_root()
        self.max_bytes = max_bytes
        self.builder = builder
        self.stale_lock_seconds = stale_lock_seconds
        self.wait_seconds = wait_seconds
        self._active: set[str] = set()
        self.root.mkdir(parents=True, exist_ok=True)
        self.cleanup()

    @staticmethod
    def _atomic_manifest_write(path: Path, manifest: dict) -> None:
        temporary = path.with_name(f".{path.name}.{os.getpid()}.{uuid.uuid4().hex}.tmp")
        try:
            with temporary.open("w", encoding="utf-8") as handle:
                json.dump(manifest, handle, ensure_ascii=False, separators=(",", ":"))
                handle.flush()
                os.fsync(handle.fileno())
            os.replace(temporary, path)
        finally:
            temporary.unlink(missing_ok=True)

    @staticmethod
    def _entry_size(entry: Path) -> int:
        return sum(item.stat().st_size for item in entry.rglob("*") if item.is_file())

    def _valid_entry(self, entry: Path, fingerprint: Optional[SourceFingerprint] = None) -> bool:
        try:
            manifest = validate_manifest(
                json.loads((entry / "manifest.json").read_text(encoding="utf-8"))
            )
            if fingerprint is not None and manifest.get("sourceFingerprint") not in {
                None,
                fingerprint.key,
            }:
                return False
            data_path = entry / manifest["dataFile"]
            with data_path.open("rb") as handle:
                return handle.read(len(DATA_MAGIC)) == DATA_MAGIC
        except (OSError, ValueError, KeyError, json.JSONDecodeError):
            return False

    def _touch_entry(self, entry: Path, fingerprint: SourceFingerprint) -> None:
        manifest_path = entry / "manifest.json"
        manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
        manifest["sourceFingerprint"] = fingerprint.key
        manifest["lastAccessNs"] = str(time.time_ns())
        self._atomic_manifest_write(manifest_path, manifest)

    def _open_fingerprint(self, fingerprint: SourceFingerprint) -> Optional[Path]:
        entry = self.root / fingerprint.key
        if not self._valid_entry(entry, fingerprint):
            return None
        self._touch_entry(entry, fingerprint)
        self._active.add(fingerprint.key)
        return entry

    def open_existing(self, source: Path) -> Optional[Path]:
        return self._open_fingerprint(source_fingerprint(source))

    def get_or_build(
        self,
        source: Path,
        *,
        on_metadata: Optional[Callable[[dict], None]] = None,
        on_progress: Optional[Callable[[dict], None]] = None,
        cancelled: Optional[Callable[[], bool]] = None,
    ) -> Path:
        source = Path(source).resolve()
        fingerprint = source_fingerprint(source)
        while True:
            existing = self._open_fingerprint(fingerprint)
            if existing is not None:
                return existing
            if cancelled is not None and cancelled():
                from src.domain.services.vcd_index_service import VcdIndexCancelled

                raise VcdIndexCancelled()
            lock = _CacheLock(
                self.root / f"{fingerprint.key}.lock",
                self.stale_lock_seconds,
            )
            if lock.acquire():
                break
            if on_progress is not None:
                on_progress({"phase": "waiting", "completed": 0, "total": 0, "percent": 0})
            time.sleep(self.wait_seconds)

        temporary = self.root / (
            f"{fingerprint.key}.tmp.{os.getpid()}.{uuid.uuid4().hex}"
        )
        final = self.root / fingerprint.key
        try:
            existing = self._open_fingerprint(fingerprint)
            if existing is not None:
                return existing
            self.builder(
                source,
                temporary,
                on_metadata=on_metadata,
                on_progress=on_progress,
                cancelled=cancelled,
            )
            self._touch_entry(temporary, fingerprint)
            if not self._valid_entry(temporary, fingerprint):
                raise ValueError("waveform cache build did not produce a valid index")
            if final.exists():
                shutil.rmtree(temporary, ignore_errors=True)
            else:
                os.replace(temporary, final)
            self._active.add(fingerprint.key)
            self.cleanup(protected={fingerprint.key})
            return final
        finally:
            shutil.rmtree(temporary, ignore_errors=True)
            lock.release()

    def release(self, index_dir: Path) -> None:
        self._active.discard(Path(index_dir).name)

    def cleanup(self, protected: Optional[set[str]] = None) -> None:
        protected_keys = self._active | set(protected or ())
        entries: list[tuple[int, int, Path]] = []
        total = 0
        for entry in self.root.iterdir():
            if not entry.is_dir() or ".tmp." in entry.name:
                continue
            if not self._valid_entry(entry):
                if entry.name not in protected_keys:
                    shutil.rmtree(entry, ignore_errors=True)
                continue
            try:
                manifest = json.loads((entry / "manifest.json").read_text(encoding="utf-8"))
                access_ns = int(manifest.get("lastAccessNs", entry.stat().st_mtime_ns))
                size = self._entry_size(entry)
            except (OSError, ValueError, json.JSONDecodeError):
                continue
            total += size
            entries.append((access_ns, size, entry))

        for _access_ns, size, entry in sorted(entries, key=lambda item: item[0]):
            if total <= self.max_bytes:
                break
            if entry.name in protected_keys:
                continue
            shutil.rmtree(entry, ignore_errors=True)
            if not entry.exists():
                total -= size

File: packages/waveform-runtime/src/waveformCache.ts
import * as crypto from 'crypto';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';

import { DATA_MAGIC, INDEX_VERSION, validateManifest } from './vcdIndexFormat';
import {
    VcdIndexBuildOptions,
    VcdIndexCancelled,
    buildVcdIndex,
} from './vcdIndex';

export const DEFAULT_WAVEFORM_CACHE_BYTES = 4 * 1024 ** 3;
const SAMPLE_BYTES = 64 * 1024;

export type SourceFingerprint = {
    key: string;
    normalizedPath: string;
    size: number;
    mtimeNs: string;
    headSha256: string;
    tailSha256: string;
    indexVersion: number;
};

export type WaveformCacheConstructorOptions = {
    root?: string;
    maxBytes?: number;
    builder?: typeof buildVcdIndex;
    staleLockMs?: number;
    waitMs?: number;
};

export function waveformCacheRoot(): string {
    if (process.platform === 'win32') {
        const base = process.env.LOCALAPPDATA ?? path.join(os.homedir(), 'AppData', 'Local');
        return path.join(base, 'VeriFlow', 'waveform-cache');
    }
    if (process.platform === 'darwin') {
        return path.join(os.homedir(), 'Library', 'Caches', 'VeriFlow', 'waveform-cache');
    }
    const base = process.env.XDG_CACHE_HOME ?? path.join(os.homedir(), '.cache');
    return path.join(base, 'veriflow', 'waveform-cache');
}

function normalizedSourcePath(source: string): string {
    const normalized = path.resolve(source).replace(/\\/g, '/');
    return process.platform === 'win32' ? normalized.toLowerCase() : normalized;
}

async function sampleHashes(source: string, size: bigint): Promise<[string, string]> {
    const handle = await fs.promises.open(source, 'r');
    try {
        const headLength = Number(size < BigInt(SAMPLE_BYTES) ? size : BigInt(SAMPLE_BYTES));
        const tailLength = headLength;
        const head = Buffer.alloc(headLength);
        const tail = Buffer.alloc(tailLength);
        await handle.read(head, 0, head.length, 0);
        await handle.read(tail, 0, tail.length, Number(size - BigInt(tailLength)));
        return [
            crypto.createHash('sha256').update(head).digest('hex'),
            crypto.createHash('sha256').update(tail).digest('hex'),
        ];
    } finally {
        await handle.close();
    }
}

export async function sourceFingerprint(source: string): Promise<SourceFingerprint> {
    source = path.resolve(source);
    const stat = await fs.promises.stat(source, { bigint: true });
    const normalizedPath = normalizedSourcePath(source);
    const [headSha256, tailSha256] = await sampleHashes(source, stat.size);
    const digest = crypto.createHash('sha256');
    digest.update(Buffer.from('VFI-CACHE-1\0', 'ascii'));
    digest.update(Buffer.from(normalizedPath, 'utf8'));
    digest.update(Buffer.from('\0', 'ascii'));
    digest.update(Buffer.from(stat.size.toString(), 'ascii'));
    digest.update(Buffer.from('\0', 'ascii'));
    digest.update(Buffer.from(stat.mtimeNs.toString(), 'ascii'));
    digest.update(Buffer.from('\0', 'ascii'));
    digest.update(Buffer.from(headSha256, 'hex'));
    digest.update(Buffer.from(tailSha256, 'hex'));
    digest.update(Buffer.from(String(INDEX_VERSION), 'ascii'));
    return {
        key: digest.digest('hex'),
        normalizedPath,
        size: Number(stat.size),
        mtimeNs: stat.mtimeNs.toString(),
        headSha256,
        tailSha256,
        indexVersion: INDEX_VERSION,
    };
}

function pidAlive(pid: number): boolean {
    if (!Number.isInteger(pid) || pid <= 0) return false;
    try {
        process.kill(pid, 0);
        return true;
    } catch (error) {
        return (error as NodeJS.ErrnoException).code === 'EPERM';
    }
}

export class WaveformCacheLock {
    private readonly token = crypto.randomBytes(16).toString('hex');
    private heartbeat: NodeJS.Timeout | undefined;
    private heartbeatWrite: Promise<void> = Promise.resolve();
    private acquired = false;

    constructor(private readonly lockPath: string, private readonly staleMs: number) {}

    private payload(): Record<string, unknown> {
        const now = Date.now();
        return { pid: process.pid, createdAtMs: now, heartbeatMs: now, token: this.token };
    }

    private async writeInitial(): Promise<void> {
        const handle = await fs.promises.open(this.lockPath, 'wx');
        try {
            await handle.writeFile(JSON.stringify(this.payload()), 'utf8');
            await handle.sync();
        } finally {
            await handle.close();
        }
    }

    private async writeHeartbeat(): Promise<void> {
        const temporary = `${this.lockPath}.heartbeat.${this.token}`;
        try {
            const current = JSON.parse(await fs.promises.readFile(this.lockPath, 'utf8'));
            if (current.token !== this.token) return;
            await fs.promises.writeFile(temporary, JSON.stringify(this.payload()), 'utf8');
            await fs.promises.rename(temporary, this.lockPath);
        } catch {
            // Another process may reclaim or remove the lock during shutdown.
        } finally {
            await fs.promises.rm(temporary, { force: true }).catch(() => undefined);
        }
    }

    private async reclaimIfStale(): Promise<boolean> {
        let original: Buffer;
        let modifiedMs: number;
        try {
            original = await fs.promises.readFile(this.lockPath);
            modifiedMs = (await fs.promises.stat(this.lockPath)).mtimeMs;
        } catch {
            return false;
        }
        let stale: boolean;
        try {
            const payload = JSON.parse(original.toString('utf8')) as Record<string, unknown>;
            const heartbeatMs = Number(payload.heartbeatMs ?? 0);
            const pid = Number(payload.pid ?? 0);
            stale = !pidAlive(pid) || Date.now() - heartbeatMs > this.staleMs;
        } catch {
            stale = Date.now() - modifiedMs > this.staleMs;
        }
        if (!stale) return false;
        try {
            const current = await fs.promises.readFile(this.lockPath);
            if (!current.equals(original)) return false;
            await fs.promises.unlink(this.lockPath);
            return true;
        } catch {
            return false;
        }
    }

    async acquire(): Promise<boolean> {
        for (let attempt = 0; attempt < 2; attempt++) {
            try {
                await this.writeInitial();
                this.acquired = true;
                this.heartbeat = setInterval(
                    () => {
                        this.heartbeatWrite = this.heartbeatWrite.then(
                            () => this.writeHeartbeat(),
                            () => this.writeHeartbeat()
                        );
                    },
                    Math.max(100, Math.floor(this.staleMs / 3))
                );
                this.heartbeat.unref();
                return true;
            } catch (error) {
                if ((error as NodeJS.ErrnoException).code !== 'EEXIST') throw error;
                if (!await this.reclaimIfStale()) return false;
            }
        }
        return false;
    }

    async release(): Promise<void> {
        if (!this.acquired) return;
        if (this.heartbeat) clearInterval(this.heartbeat);
        await this.heartbeatWrite;
        try {
            const payload = JSON.parse(await fs.promises.readFile(this.lockPath, 'utf8'));
            if (payload.token === this.token) await fs.promises.unlink(this.lockPath);
        } catch {
            // The lock may already have been reclaimed after an owner failure.
        }
        this.acquired = false;
    }
}

export class WaveformCache {
    public readonly root: string;
    public maxBytes: number;
    private readonly builder: typeof buildVcdIndex;
    private readonly staleLockMs: number;
    private readonly waitMs: number;
    private readonly active = new Set<string>();
    private readonly startupCleanup: Promise<void>;

    constructor(options: WaveformCacheConstructorOptions = {}) {
        this.root = path.resolve(options.root ?? waveformCacheRoot());
        this.maxBytes = options.maxBytes ?? DEFAULT_WAVEFORM_CACHE_BYTES;
        if (!Number.isFinite(this.maxBytes) || this.maxBytes <= 0) {
            throw new Error('waveform cache size must be positive');
        }
        this.builder = options.builder ?? buildVcdIndex;
        this.staleLockMs = options.staleLockMs ?? 30_000;
        this.waitMs = options.waitMs ?? 100;
        fs.mkdirSync(this.root, { recursive: true });
        this.startupCleanup = this.cleanupInternal();
    }

    private async validEntry(entry: string, fingerprint?: SourceFingerprint): Promise<boolean> {
        try {
            const manifest = validateManifest(
                JSON.parse(await fs.promises.readFile(path.join(entry, 'manifest.json'), 'utf8'))
            ) as Record<string, unknown>;
            if (
                fingerprint &&
                manifest.sourceFingerprint !== undefined &&
                manifest.sourceFingerprint !== fingerprint.key
            ) return false;
            const data = await fs.promises.open(path.join(entry, manifest.dataFile as string), 'r');
            try {
                const magic = Buffer.alloc(DATA_MAGIC.length);
                const { bytesRead } = await data.read(magic, 0, magic.length, 0);
                return bytesRead === magic.length && magic.equals(DATA_MAGIC);
            } finally {
                await data.close();
            }
        } catch {
            return false;
        }
    }

    private async atomicManifestWrite(manifestPath: string, manifest: unknown): Promise<void> {
        const temporary = path.join(
            path.dirname(manifestPath),
            `.${path.basename(manifestPath)}.${process.pid}.${crypto.randomBytes(8).toString('hex')}.tmp`
        );
        try {
            const handle = await fs.promises.open(temporary, 'w');
            try {
                await handle.writeFile(JSON.stringify(manifest), 'utf8');
                await handle.sync();
            } finally {
                await handle.close();
            }
            await fs.promises.rename(temporary, manifestPath);
        } finally {
            await fs.promises.rm(temporary, { force: true }).catch(() => undefined);
        }
    }

    private async touchEntry(entry: string, fingerprint: SourceFingerprint): Promise<void> {
        const manifestPath = path.join(entry, 'manifest.json');
        const manifest = JSON.parse(await fs.promises.readFile(manifestPath, 'utf8'));
        manifest.sourceFingerprint = fingerprint.key;
        manifest.lastAccessNs = (BigInt(Date.now()) * 1_000_000n).toString();
        await this.atomicManifestWrite(manifestPath, manifest);
    }

    private async openFingerprint(fingerprint: SourceFingerprint): Promise<string | null> {
        const entry = path.join(this.root, fingerprint.key);
        if (!await this.validEntry(entry, fingerprint)) {
            if (fs.existsSync(entry) && !this.active.has(fingerprint.key)) {
                await fs.promises.rm(entry, { recursive: true, force: true });
            }
            return null;
        }
        await this.touchEntry(entry, fingerprint);
        this.active.add(fingerprint.key);
        return entry;
    }

    async openExisting(source: string): Promise<string | null> {
        await this.startupCleanup;
        return this.openFingerprint(await sourceFingerprint(source));
    }

    async getOrBuild(
        source: string,
        options: VcdIndexBuildOptions = {}
    ): Promise<string> {
        await this.startupCleanup;
        source = path.resolve(source);
        const fingerprint = await sourceFingerprint(source);
        let lock: WaveformCacheLock;
        while (true) {
            const existing = await this.openFingerprint(fingerprint);
            if (existing) return existing;
            if (options.cancelled?.()) throw new VcdIndexCancelled();
            lock = new WaveformCacheLock(
                path.join(this.root, `${fingerprint.key}.lock`),
                this.staleLockMs
            );
            if (await lock.acquire()) break;
            options.onProgress?.({ phase: 'waiting', completed: 0, total: 0, percent: 0 });
            await new Promise(resolve => setTimeout(resolve, this.waitMs));
        }

        const temporary = path.join(
            this.root,
            `${fingerprint.key}.tmp.${process.pid}.${crypto.randomBytes(8).toString('hex')}`
        );
        const final = path.join(this.root, fingerprint.key);
        try {
            const existing = await this.openFingerprint(fingerprint);
            if (existing) return existing;
            await this.builder(source, temporary, options);
            await this.touchEntry(temporary, fingerprint);
            if (!await this.validEntry(temporary, fingerprint)) {
                throw new Error('waveform cache build did not produce a valid index');
            }
            if (fs.existsSync(final)) {
                await fs.promises.rm(temporary, { recursive: true, force: true });
            } else {
                await fs.promises.rename(temporary, final);
            }
            this.active.add(fingerprint.key);
            await this.cleanupInternal(new Set([fingerprint.key]));
            return final;
        } finally {
            await fs.promises.rm(temporary, { recursive: true, force: true }).catch(() => undefined);
            await lock.release();
        }
    }

    release(indexDir: string): void {
        this.active.delete(path.basename(path.resolve(indexDir)));
    }

    private async directorySize(directory: string): Promise<number> {
        let total = 0;
        for (const entry of await fs.promises.readdir(directory, { withFileTypes: true })) {
            const item = path.join(directory, entry.name);
            if (entry.isDirectory()) total += await this.directorySize(item);
            else if (entry.isFile()) total += (await fs.promises.stat(item)).size;
        }
        return total;
    }

    private async cleanupInternal(protectedKeys = new Set<string>()): Promise<void> {
        const protectedEntries = new Set([...this.active, ...protectedKeys]);
        const entries: Array<{ accessNs: bigint; size: number; entry: string; key: string }> = [];
        let total = 0;
        for (const directoryEntry of await fs.promises.readdir(this.root, { withFileTypes: true })) {
            if (!directoryEntry.isDirectory() || directoryEntry.name.includes('.tmp.')) continue;
            const entry = path.join(this.root, directoryEntry.name);
            if (!await this.validEntry(entry)) {
                if (!protectedEntries.has(directoryEntry.name)) {
                    await fs.promises.rm(entry, { recursive: true, force: true });
                }
                continue;
            }
            try {
                const manifest = JSON.parse(
                    await fs.promises.readFile(path.join(entry, 'manifest.json'), 'utf8')
                );
                const stat = await fs.promises.stat(entry, { bigint: true });
                const accessNs = BigInt(manifest.lastAccessNs ?? stat.mtimeNs);
                const size = await this.directorySize(entry);
                total += size;
                entries.push({ accessNs, size, entry, key: directoryEntry.name });
            } catch {
                // A concurrent publisher or accessor may be replacing the manifest.
            }
        }
        entries.sort((left, right) => left.accessNs < right.accessNs ? -1 : left.accessNs > right.accessNs ? 1 : 0);
        for (const item of entries) {
            if (total <= this.maxBytes) break;
            if (protectedEntries.has(item.key)) continue;
            await fs.promises.rm(item.entry, { recursive: true, force: true });
            total -= item.size;
        }
    }

    async cleanup(): Promise<void> {
        await this.startupCleanup;
        await this.cleanupInternal();
    }
}
